aip/cloud/25146.md

```diff
@@ -0,0 +1,25 @@
+---
+id: 25146
+state: draft
+created: TBD
+placement:
+  category: gcp
+---
+
+# Generic Fields
+
+Guidelines related to generic fields in [AIP-146](../0146.md) **should** be applied
+with the following exceptions.
+
+## Maps
+
+Fields of type `map` **must not** be composed of type Message. `map<primitive, primitive>`
+is permissible, but `map<primitive, google.protobuf.Struct>` is not allowed.
```

Comment:
Clarification: Is `map<primitive, Book>` allowed i.e. well-defined messages can be the value type, but "generic" messages cannot? Or is it strictly all `message` types?

Reply:
All message types.

When the key is dynamic we lose the clearly defined structure that is needed for declarative surfaces

Reply:
> When the key is dynamic we lose the clearly defined structure that is needed for declarative surfaces

I don't follow.

What is not "clearly defined" about `map<string, Book>`? Clients should know exactly what the `Book` looks like.

Reply:
The problem is that clients often can't construct a schema with an arbitrary key that maps to a complex object.

This is a restriction in Terraform as well as KCC, the schema must be known at compile time **including** any keys except for simple string -> primitive maps

Basically you have to give Terraform the full schema of an object you want to represent where every settable field must have a name known at compile time. Dynamic keys where the value is an object cannot be modeled without changes.

Reply:
Ok, I see, there is a pretty strict technological limitation here beyond our own tooling's control. Let me do a quick check on how broadly this effects existing APIs.

Reply:
Ok so, there are over 200 fields that would violate this AIP in `google/cloud` alone. Of those, 45 of them belong to messages that define a resource. The specific fields are logged in text files attached to this comment. These were generated using a little tool I just wrote, code is here: https://github.com/noahdietz/messagemaps.

How are those fields being handled today, if at all?

[resources_only.txt](https://github.com/aip-dev/google.aip.dev/files/10025841/resources_only.txt)
[all.txt](https://github.com/aip-dev/google.aip.dev/files/10025842/all.txt)

Reply:
Often we don't support these resources. If we need to we have to get creative with how we support it.

For example, GKEHub Feature has a MembershipFeatureSpec. We model this as a new resource: https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/gke_hub_feature_membership where the `membership` field in the gke_hub_feature_membership resource is the key in the map

Another example is InstanceGroupManager's perInstanceConfig field: https://cloud.google.com/compute/docs/reference/rest/v1/instanceGroupManagers/patchPerInstanceConfigs where the disks field is 

`Preserved disks defined for this instance. This map is keyed with the device names of the disks.`

We model this in Terraform as https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_per_instance_config where the device name of the disk is a normal field in an object, and we translate that to the required format when sending the patch request

Reply:
Given the prevalence of such fields and the existing guidance that supports use of `map<string, Message>` in my other comment, I think we need to take a long look at the technical solutions for what are technical problems. The AIPs should not aim to solve one technology's (potentially fixable) shortcomings that in turn limits valid API design patterns. As such, I'd like to see a design to fix the issues mentioned that has been rejected before moving forward with this suggested guidance.

> Ok, I see, there is a pretty strict technological limitation here beyond our own tooling's control

I said this earlier. I realize that this may necessitate contributions to the core technology that aren't solely owned by Google. Presumably, the technologies afflicted by the presented friction would appreciate expanding support for more API surfaces.

FYI @shwoodard and @toumorokoshi

Reply:
sorry just catching up, and skimmed a first pass.

My high-level perspective is that we should push toward patterns that help enable both API usability which should, in turn, enable client surface usability. At minimum it should be beneficial for clients, neutral for APIs.

At first glance not allowing map<string, Message> doesn't match that criteria, IMO: I worry it will force users to effectively normalize the resource schema into some primitive (likely a string), when it is possible to have a more structured schema.

Also most clients support some form of nested message schema: this isn't a general client surface usability improvement, this would be quite specific for Terraform.

If a client surface is deficient in expressing API behavior, we should explore both if the API could be improved, and also if the client surface can have some pattern to work around said defiiciency.

I'll try to read a bit more deeply and leave some other thoughts too.